crates/turbopack-core/src/resolve/mod.rs

```diff
@@ -493,29 +598,53 @@ impl ResolveResult {
             primary: self
                 .primary
                 .iter()
-                .cloned()
-                .map(|item| {
+                .map(|(request, item)| {
                     let asset_fn = &source_fn;
+                    let request = request.clone();
+                    let item = item.clone();
                     async move {
-                        Ok(match item {
-                            ResolveResultItem::Source(source) => asset_fn(source).await?,
-                            ResolveResultItem::OriginalReferenceTypeExternal(s) => {
-                                ModuleResolveResultItem::OriginalReferenceTypeExternal(s)
-                            }
-                            ResolveResultItem::Ignore => ModuleResolveResultItem::Ignore,
-                            ResolveResultItem::Empty => ModuleResolveResultItem::Empty,
-                            ResolveResultItem::Custom(u8) => ModuleResolveResultItem::Custom(u8),
-                            ResolveResultItem::Unresolveable => {
-                                ModuleResolveResultItem::Unresolveable
-                            }
-                        })
+                        Ok((
+                            request,
+                            match item {
+                                ResolveResultItem::Source(source) => asset_fn(source).await?,
+                                ResolveResultItem::OriginalReferenceTypeExternal(s) => {
+                                    ModuleResolveResultItem::OriginalReferenceTypeExternal(s)
+                                }
+                                ResolveResultItem::Ignore => ModuleResolveResultItem::Ignore,
+                                ResolveResultItem::Empty => ModuleResolveResultItem::Empty,
+                                ResolveResultItem::Custom(u8) => {
+                                    ModuleResolveResultItem::Custom(u8)
+                                }
+                                ResolveResultItem::Unresolveable => {
+                                    ModuleResolveResultItem::Unresolveable
+                                }
+                            },
+                        ))
                     }
                 })
                 .try_join()
-                .await?,
+                .await?
+                .into_iter()
+                .collect(),
             affecting_sources: self.affecting_sources.clone(),
         })
     }
+
+    pub fn with_request_ref(&self, request: String) -> Self {
```

Comment:
```suggestion
    pub fn with_request_key_ref(&self, request_key: String) -> Self {
```

Also, why the `_ref` in the name?

Reply:
Being pedantic as just having "request" makes it sound like it's the full `Request` that was passed to `resolve(...)`

Reply:
> Also, why the _ref in the name?

alex started that pattern when the same method name is defined on Vc<X> and X...